File: lib/services/anime_service.dart
import 'dart:convert';
import 'package:html/parser.dart';
import 'package:movi/models/anime_search_result_model.dart';
import 'package:movi/services/networking.dart';

class AnimeService {
  static final _baseUrl = 'https://www.kickassanime.rs';

  static String _queryUrlConstructor(String query) {
    return '$_baseUrl/search?q=$query';
  }

  static Future<List<AnimeSearchResultModel>> searchAnime(String query) async {
    List<AnimeSearchResultModel> l = [];

    var url = _queryUrlConstructor(query);
    var responseBody = await Networking.getStringResponse(url);

    if (responseBody == Networking.ERROR) {
      return l;
    }

    var document = parse(responseBody);
    var scripts = document.querySelectorAll('script');
    var rawScript = '';
    for (var s in scripts) {
      if (s.text.contains('appData')) {
        // script contains the animeList
        rawScript = s.text;
      }
    }

    var rawJson = rawScript.split(' appData = ')[1].split(' || {}')[0];
    var jsonAnimeList = jsonDecode(rawJson)['animes'];

    for (var anime in jsonAnimeList) {
      l.add(AnimeSearchResultModel(
        animeUrl: _baseUrl + anime['slug'].toString().replaceAll('\\', ''),
        posterUrl:
            anime['image'].toString().replaceAll('\\', '') + anime['poster'],
        titleText: anime['name'],
      ));
    }

    return l;
  }
}

File: movi-api/services/animeService.js
const { parse } = require("node-html-parser");

const Networking = require("./networking");
const StreamService = require("./streamService");

const baseUrl = "https://www1.kickassanime.rs/";

class AnimeService {
  // method for converting html body to appData json object
  static getRawJson(responseBody) {
    const root = parse(responseBody, {
      script: true,
    });
    const scripts = root.querySelectorAll("script");

    const appDataScript = scripts.find((script) =>
      script.rawText.includes("appData")
    );

    const jsonResponse = appDataScript
      .toString()
      .split(" appData = ")[1]
      .split(" || {}")[0];

    return JSON.parse(jsonResponse);
  }

  static queryAnimeBy(searchedQuery) {
    const url = `${baseUrl}search?q=${searchedQuery}`;

    return new Promise((resolve, reject) => {
      Networking.getResponse(url, true)
        .then((response) => resolve(this.getRawJson(response).animes))
        .catch((err) => reject(err));
    });
  }

  static queryAnimeDetails(animeID) {
    const url = `${baseUrl}/anime/${animeID}`;

    return new Promise((resolve, reject) => {
      Networking.getResponse(url, true)
        .then((response) => resolve(this.getRawJson(response).anime))
        .catch((err) => reject(err));
    });
  }

  static getJsonObjectWithPlayableLink(jsonResponse, link) {
    return {
      anime: jsonResponse.anime,
      episode: jsonResponse.episode,
      playableLink: link,
      episodes: jsonResponse.episodes,
    };
  }

  static queryEpisodeDetails(animeID, episodeID) {
    const url = `${baseUrl}/anime/${animeID}/${episodeID}`;

    return new Promise((resolve, reject) => {
      Networking.getResponse(url, true)
        .then((response) => {
          const jsonResponse = this.getRawJson(response);

          try {
            // try finding for external servers
            const extServers = jsonResponse.ext_servers;
            const vidstreamingService = extServers.find(
              (s) => s.name === "Vidstreaming"
            );

            StreamService.getPlayableUrl(vidstreamingService.link)
              .then((link) =>
                resolve(this.getJsonObjectWithPlayableLink(jsonResponse, link))
              )
              .catch((_) =>
                StreamService.getBackupPlayableUrl(jsonResponse)
                  .then((link) =>
                    resolve(
                      this.getJsonObjectWithPlayableLink(jsonResponse, link)
                    )
                  )
                  .catch((_) =>
                    resolve(
                      this.getJsonObjectWithPlayableLink(jsonResponse, null)
                    )
                  )
              );
          } catch (_) {
            StreamService.getBackupPlayableUrl(jsonResponse)
              .then((link) =>
                resolve(this.getJsonObjectWithPlayableLink(jsonResponse, link))
              )
              .catch((_) =>
                resolve(this.getJsonObjectWithPlayableLink(jsonResponse, null))
              );
          }
        })
        .catch((err) => reject(err));
    });
  }
}

module.exports = AnimeService;
